x/yarpctest/request_stream.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpctest
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/transport/grpc"
+	"go.uber.org/yarpc/x/yarpctest/api"
+)
+
+// GRPCStreamRequest creates a new grpc stream request.
+func GRPCStreamRequest(options ...api.ClientStreamRequestOption) api.Action {
+	opts := api.NewClientStreamRequestOpts()
+	for _, option := range options {
+		option.ApplyClientStreamRequest(&opts)
+	}
+	return api.ActionFunc(func(t testing.TB) {
+		trans := grpc.NewTransport()
+		out := trans.NewSingleOutbound(fmt.Sprintf("127.0.0.1:%d", opts.Port))
+
+		require.NoError(t, trans.Start())
+		defer func() { assert.NoError(t, trans.Stop()) }()
+
+		require.NoError(t, out.Start())
+		defer func() { assert.NoError(t, out.Stop()) }()
+
+		res, err := callStream(t, out, opts.GiveRequest, opts.StreamActions)
+		if len(opts.WantErrMsgs) > 0 {
+			require.Error(t, err)
+			for _, wantErrMsg := range opts.WantErrMsgs {
+				require.Contains(t, err.Error(), wantErrMsg)
```

Comment:
Are `api.Action`s guaranteed to run from the test goroutine? `require` will
call `t.FailNow` in case of a failure, which will kill the "current"
goroutine. If this ever gets called from the goroutine handling the request,
the request will abruptly stop being handled and the client will just see a
timeout.

Reply:
Actions are currently sequential/in the test goroutine.  My assumption was that the testing.T would get annotated with the error information regardless, this is possibly something that can get fixed in a follow-up